backend/src/model/entity/plant_layer.rs

```diff
@@ -12,10 +17,91 @@ use crate::{
 };
 
+/// The resolution of the generated heatmap in cm.
+const GRANULARITY: f32 = 10.0;
+
+/// A bounding box around the maps geometry.
+#[derive(Debug, Clone, QueryableByName)]
+struct BoundingBox {
+    /// The lowest x value in the geometry.
+    #[diesel(sql_type = Float)]
+    x_min: f32,
+    /// The lowest y value in the geometry.
+    #[diesel(sql_type = Float)]
+    y_min: f32,
+    /// The highest x value in the geometry.
+    #[diesel(sql_type = Float)]
+    x_max: f32,
+    /// The highest y value in the geometry.
+    #[diesel(sql_type = Float)]
+    y_max: f32,
+}
+
+/// Stores the score of a x,y coordinate on the heatmap.
+#[derive(Debug, Clone, QueryableByName)]
+struct HeatMapElement {
+    /// The score on the heatmap.
+    #[diesel(sql_type = Float)]
+    score: f32,
+    /// The x values of the score
+    #[diesel(sql_type = Integer)]
+    x: i32,
+    /// The y values of the score.
+    #[diesel(sql_type = Integer)]
+    y: i32,
+}
+
+/// Generates a heatmap signaling ideal locations for planting the plant.
+///
+/// # Errors
+/// * If the SQL query failed.
+#[allow(
+    clippy::cast_sign_loss,
+    clippy::indexing_slicing,
+    clippy::cast_possible_truncation    // ok, because ceil prevents invalid truncation
+)]
+pub async fn heatmap(
+    map_id: i32,
+    layer_id: i32,
+    plant_id: i32,
+    conn: &mut AsyncPgConnection,
+) -> QueryResult<Vec<Vec<f32>>> {
+    // Fetch the bounding box x and y values of the maps coordinates
+    let bounding_box_query =
+        diesel::sql_query("SELECT * FROM calculate_bbox($1)").bind::<Integer, _>(map_id);
+    debug!("{}", debug_query::<Pg, _>(&bounding_box_query));
+    let bounding_box = bounding_box_query.get_result::<BoundingBox>(conn).await?;
+
+    // Fetch the heatmap
+    let query = diesel::sql_query("SELECT * FROM calculate_score($1, $2, $3, $4, $5, $6, $7, $8)")
+        .bind::<Integer, _>(map_id)
+        .bind::<Integer, _>(layer_id)
+        .bind::<Integer, _>(plant_id)
+        .bind::<Float, _>(GRANULARITY)
+        .bind::<Float, _>(bounding_box.x_min)
+        .bind::<Float, _>(bounding_box.y_min)
+        .bind::<Float, _>(bounding_box.x_max)
+        .bind::<Float, _>(bounding_box.y_max);
+    debug!("{}", debug_query::<Pg, _>(&query));
+    let result = query.load::<HeatMapElement>(conn).await?;
+
+    // Convert the result to a matrix.
+    // Matrix will be from 0..0 to ((x_max - x_min) / granularity)..((y_max - y_min) / granularity).
+    let mut heatmap =
+        vec![
+            vec![0.0; ((bounding_box.x_max - bounding_box.x_min) / GRANULARITY).ceil() as usize];
+            ((bounding_box.y_max - bounding_box.y_min) / GRANULARITY).ceil() as usize
+        ];
+    for HeatMapElement { score, x, y } in result {
+        heatmap[y as usize][x as usize] = score;
+    }
+    Ok(heatmap)
+}
+
 /// Get all relations of a certain plant.
 ///
 /// # Errors
-/// * Unknown, diesel doesn't say why it might error.
+/// * If the SQL query failed.
```

Comment:
I'm not sure, but it might actually return a 404 if there are no rows returned. I don't know exactly how diesel handles that, but in some instances it returns 404 on no rows found.